File: python_sdk/src/reality_capture/specifications/objects2d.py
from pydantic import BaseModel, Field
from typing import Optional
from enum import Enum


class Objects2DInputs(BaseModel):
    photos: str = Field(description="Reality data id of ContextScene, pointing to photos to process")
    photo_object_detector: Optional[str] = Field(None, alias="photoObjectDetector",
                                                 description="Either reality data id of photo object detector "
                                                             "or photo object detector identifier "
                                                             "from the AI Detectors library")
    point_clouds: Optional[str] = Field(None, alias="pointClouds",
                                        description="Reality data id of ContextScene, "
                                                    "pointing to a collection of point clouds to process")
    meshes: Optional[str] = Field(None,
                                  description="Reality data id of ContextScene, "
                                              "pointing to a collection of meshes to process")
    objects2d: Optional[str] = Field(None, alias="objects2D",
                                     description="Reality data id of ContextScene, annotated with embedded 2D objects, "
                                                 "this input replaces photo_object_detector input")


class Objects2DOutputs(BaseModel):
    objects2d: Optional[str] = Field(None, alias="objects2D",
                                     description="Reality data id of ContextScene, annotated with embedded 2D objects")
    objects3d: Optional[str] = Field(None, alias="objects3D",
                                     description="Reality data id of ContextScene, annotated with embedded 3D objects")
    objects3d_as_dgn: Optional[str] = Field(None, alias="objects3DAsDGN",
                                            description="Reality data id of 3D objects as DGN file, "
                                                        "objects3d output must be defined")
    objects3d_as_3d_tiles: Optional[str] = Field(None, alias="objects3DAs3DTiles",
                                                 description="Reality data id of 3D objects as 3D Tiles file, "
                                                             "objects3d output must be defined")
    objects3d_as_geojson: Optional[str] = Field(None, alias="objects3DAsGeoJSON",
                                                description="Reality data id of 3D objects as GeoJSON file, "
                                                            "objects3d output must be defined")
    locations3d_as_shp: Optional[str] = Field(None, alias="locations3DAsSHP",
                                              description="Reality data id of 3D objects locations as SHP file, "
                                                          "objects3d output must be defined")
    locations3d_as_geojson: Optional[str] = Field(None, alias="locations3DAsGeoJSON",
                                                  description="Reality data id of 3D objects locations "
                                                              "as GeoJSON file, "
                                                              "objects3d output must be defined")


class Objects2DOutputsCreate(Enum):
    OBJECTS2D = "objects2D"
    OBJECTS3D = "objects3D"
    OBJECTS3D_AS_DGN = "objects3DAsDGN"
    """
    Requires OBJECTS3D in order to be produced.
    """
    OBJECTS3D_AS_3DTILES = "objects3DAs3DTiles"
    """
    Requires OBJECTS3D in order to be produced.
    """
    OBJECTS3D_AS_GEOJSON = "objects3DAsGeoJSON"
    """
    Requires OBJECTS3D in order to be produced.
    """
    LOCATIONS3D_AS_SHP = "locations3DAsSHP"
    """
    Requires OBJECTS3D in order to be produced.
    """
    LOCATIONS3D_AS_GEOJSON = "locations3DAsGeoJSON"
    """
    Requires OBJECTS3D in order to be produced.
    """


class Objects2DOptions(BaseModel):
    use_tie_points: Optional[bool] = Field(None, alias="useTiePoints",
                                           description="Improve detection using tie points in oriented photos.")
    max_dist: Optional[float] = Field(None, alias="maxDist",
                                      description="Maximum distance between photos and 3D objects")
    crs: Optional[str] = Field(None, description="CRS used by ``objects3d_as_dgn`` and ``locations3d_as_shp`` outputs")
    min_photos: Optional[int] = Field(None, alias="minPhotos",
                                      description="Minimum number of 2D objects to generate a 3D object")


class Objects2DSpecificationsCreate(BaseModel):
    inputs: Objects2DInputs = Field(description="Inputs")
    outputs: list[Objects2DOutputsCreate] = Field(description="Outputs")
    options: Optional[Objects2DOptions] = Field(None, description="Options")


class Objects2DSpecifications(BaseModel):
    inputs: Objects2DInputs = Field(description="Inputs")
    outputs: Objects2DOutputs = Field(description="Outputs")
    options: Optional[Objects2DOptions] = Field(None, description="Options")

File: typescript_sdk/reality_capture/src/specifications/objects2d.ts
/*import { z } from "zod";

export const Objects2DInputsSchema = z.object({
  photos: z.string().describe("Reality data id of ContextScene, pointing to photos to process"),
  photoObjectDetector: z.string().optional().describe(
    "Either reality data id of photo object detector or photo object detector identifier from the AI Detectors library"
  ),
  model3D: z.string().optional().describe(
    "Reality data id of ContextScene, pointing to a collection of point clouds/meshes to process, or a point cloud, or a mesh."
  ),
  objects2D: z.string().optional().describe(
    "Reality data id of ContextScene, annotated with embedded 2D objects, this input replaces photoObjectDetector input"
  ),
});
export type Objects2DInputs = z.infer<typeof Objects2DInputsSchema>;

export const Objects2DOutputsSchema = z.object({
  objects2D: z.string().optional().describe(
    "Reality data id of ContextScene, annotated with embedded 2D objects"
  ),
  objects3D: z.string().optional().describe(
    "Reality data id of ContextScene, annotated with embedded 3D objects"
  ),
  objects3DAs3DTiles: z.string().optional().describe(
    "Reality data id of 3D objects as 3D Tiles file, objects3d output must be defined"
  ),
  objects3DAsGeoJSON: z.string().optional().describe(
    "Reality data id of 3D objects as GeoJSON file, objects3d output must be defined"
  ),
  locations3DAsSHP: z.string().optional().describe(
    "Reality data id of 3D objects locations as SHP file, objects3d output must be defined"
  ),
  locations3DAsGeoJSON: z.string().optional().describe(
    "Reality data id of 3D objects locations as GeoJSON file, objects3d output must be defined"
  ),
});
export type Objects2DOutputs = z.infer<typeof Objects2DOutputsSchema>;

export enum Objects2DOutputsCreate {
  OBJECTS2D = "objects2D",
  OBJECTS3D = "objects3D",
  OBJECTS3D_AS_3DTILES = "objects3DAs3DTiles",
  OBJECTS3D_AS_GEOJSON = "objects3DAsGeoJSON",
  LOCATIONS3D_AS_SHP = "locations3DAsSHP",
  LOCATIONS3D_AS_GEOJSON = "locations3DAsGeoJSON"
}

export const Objects2DOptionsSchema = z.object({
  useTiePoints: z.boolean().optional().describe(
    "Improve detection using tie points in oriented photos."
  ),
  maxDist: z.number().optional().describe(
    "Maximum distance between photos and 3D objects"
  ),
  crs: z.string().optional().describe(
    "CRS used by ``locations3d_as_shp`` outputs"
  ),
  minPhotos: z.number().optional().describe(
    "Minimum number of 2D objects to generate a 3D object"
  ),
});
export type Objects2DOptions = z.infer<typeof Objects2DOptionsSchema>;

export const Objects2DSpecificationsCreateSchema = z.object({
  inputs: Objects2DInputsSchema.describe("Inputs"),
  outputs: z.array(z.nativeEnum(Objects2DOutputsCreate)).describe("Outputs"),
  options: Objects2DOptionsSchema.optional().describe("Options"),
});
export type Objects2DSpecificationsCreate = z.infer<typeof Objects2DSpecificationsCreateSchema>;

export const Objects2DSpecificationsSchema = z.object({
  inputs: Objects2DInputsSchema.describe("Inputs"),
  outputs: Objects2DOutputsSchema.describe("Outputs"),
  options: Objects2DOptionsSchema.optional().describe("Options"),
});
export type Objects2DSpecifications = z.infer<typeof Objects2DSpecificationsSchema>;*/